gym_handeye/handeye.py

```diff
@@ -0,0 +1,310 @@
+        :param y_end: End gripper y position
+        :return:
+        """
+        x_start = self.grip_pos_x
+        y_start = self.grip_pos_y
+
+        self.grip_pos_x = x_end
+        self.grip_pos_y = y_end
+
+        if self.block_in_hand:
+            self.block_pos_x = x_end
+            self.block_pos_y = y_end
+            self.observation[y_start * self.grid_size + x_start] = 'w'
+            self.observation[y_end * self.grid_size + x_end] = 'b'
+        else:
+            self.observation[y_start * self.grid_size + x_start] = 'w'
+            self.observation[y_end * self.grid_size + x_end] = 'g'
+
+        if self.block_pos_x == x_start and self.block_pos_y == y_start:
+            self.observation[y_start * self.grid_size + x_start] = 'b'
+
+        self.observation[self.env_size - 1] = '0'
+
+        if self.block_pos_x == x_end and self.block_pos_y == y_end:
+            if not self.block_in_hand:
+                self.observation[self.env_size - 1] = '1'
+            else:
+                if self.note_in_hand:
+                    self.observation[self.env_size - 1] = '2'
+
+        return
+
+    def _grip_block(self):
+        """
+        Executes a gripping action.
+        :return:
+        """
+        if self.block_in_hand:
+            return
+
+        if self.block_pos_x == self.grip_pos_x and self.block_pos_y == self.grip_pos_y:
+            self.block_in_hand = True
+            self.observation[self.grip_pos_y * self.grid_size + self.grip_pos_x] = 'b'
+            if self.note_in_hand:
+                self.observation[self.env_size - 1] = '2'
+            else:
+                self.observation[self.env_size - 1] = '0'
+
+        return
+
+    def _release_block(self):
+        """
+        Releases a block if a block is held in the hand.
+        :return:
+        """
+        if self.block_in_hand:
+            self.observation[self.env_size - 1] = '1'
+            self.observation[self.grip_pos_y * self.grid_size + self.grip_pos_x] = 'g'
+            self.block_in_hand = False
+        return
+
+    def _observe(self):
+        return self.observation
+
+    def _take_action(self, action, observation):
+        action_type = ACTION_LOOKUP[action]
+
+        if action_type == "N" and self.grip_pos_y > 0:
+            self._move_gripper(self.grip_pos_x, self.grip_pos_y - 1)
+
+        elif action_type == "E" and self.grip_pos_x < self.grid_size - 1:
+            self._move_gripper(self.grip_pos_x + 1, self.grip_pos_y)
+
+        elif action_type == "S" and self.grip_pos_y < self.grid_size - 1:
+            self._move_gripper(self.grip_pos_x, self.grip_pos_y + 1)
+
+        elif action_type == "W" and self.grip_pos_x > 0:
+            self._move_gripper(self.grip_pos_x - 1, self.grip_pos_y)
+
+        elif action_type == "G":
+            self._grip_block()
+
+        elif action_type == "R":
+            self._release_block()
+
+        return
+
+    def _random_positions(self):
```

Comment:
Przydałby się docstring informujący że nie jest to deterministtyczna metoda i może zwracać 2  różne rodzaje stanów.